### PR TITLE
Font subset

### PR DESCRIPTION
**Pull request by dnfield**
Adds a font subsetting utillity to the engine build.

This is meant to be consumed by the flutter tool to help with glyph shaking.  Once it lands in tree we can update the recipes to upload it to storage along with the other host artifacts.

Tagged a bunch of people for review:

@zanderso and @jonahwilliams for evaluation of how this will be used by the tool

@jason-simmons @GaryQian @chinmaygarde  because they probably know more about harfbuzz than I do, and general engine review.

@willlarche FYI (and feel free to give feedback as well).

### CODE REVIEW

**Comment by GaryQian** [comment]
Is there an issue this fixes for more context?

**Comment by dnfield** [comment]
Fixes https://github.com/flutter/flutter/issues/43642

**Comment by dnfield** [comment]
I was expecting this to run the tests I wrote on all platforms, but it looks like not unfortuantely.  That will require a recipe change, or that I make these tests run under the "engine" category for now.  I'm going to push something up that does that at least temporarily to try to get them to run on Windows, since I haven't tested it there yet.

**Comment by jonahwilliams** [comment]
Seems like this would be fairly straightforward to incorporate into the asset build, provided that we already knew the subset of fonts needed? I'm assuming that later part is where the kernel parsing comes in

**Comment by dnfield** [comment]
> Seems like this would be fairly straightforward to incorporate into the asset build, provided that we already knew the subset of fonts needed? I'm assuming that later part is where the kernel parsing comes in

Yes

**Comment by dnfield** [comment]
Having issues with the test.  I'll look at it some more later.  The script seems to try to compare before the process has finished writing files

**Comment by dnfield** [comment]
CI should be passing now.

File: tools/font-subset/main.cc [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  HarfbuzzWrappers::HbSubsetInputPtr input(hb_subset_input_create_or_fail());
+  {
+    hb_set_t* desired_codepoints = hb_subset_input_unicode_set(input.get());
+    HarfbuzzWrappers::HbSetPtr actual_codepoints(hb_set_create());
+    hb_face_collect_unicodes(font_face.get(), actual_codepoints.get());
+    for (int i = 3; i < argc; i++) {
+      auto codepoint = ParseCodepoint(argv[i]);
+      if (codepoint) {
```
Comment (zanderso) [inline tools/font-subset/main.cc]: nit:
```
if (!codepoint) {
  continue;
}
```
then unindent.

File: tools/font-subset/main.cc [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+              << "; aborting." << std::endl;
+    return -1;
+  }
+
+  HarfbuzzWrappers::HbSubsetInputPtr input(hb_subset_input_create_or_fail());
+  {
+    hb_set_t* desired_codepoints = hb_subset_input_unicode_set(input.get());
+    HarfbuzzWrappers::HbSetPtr actual_codepoints(hb_set_create());
+    hb_face_collect_unicodes(font_face.get(), actual_codepoints.get());
+    for (int i = 3; i < argc; i++) {
```
Comment (zanderso) [inline tools/font-subset/main.cc]: Is there any danger of running into limits on the number/length of command line args?

Comment (dnfield) [reply tools/font-subset/main.cc]: ...yes. Windows in particular I think is the most limited, which would limit us to either 2047 or 8191 characters.  That is likely a problem.

I'll re-work this to take an input file instead, unless you have a better idea.

Comment (zanderso) [reply tools/font-subset/main.cc]: input file sgtm

Comment (dnfield) [reply tools/font-subset/main.cc]: Actually, the way hb-subset does this is to just take the first line of stdin as a string.  I thought it was awkward to do it that way, but it seems better.  Will look at that instead.

Comment (dnfield) [reply tools/font-subset/main.cc]: Pushed up changes to read from stdin and amended tests.

File: tools/font-subset/main.cc (hunk 1 above)
Comment (dnfield) [reply tools/font-subset/main.cc]: Done

File: tools/font-subset/main.cc [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+  HarfbuzzWrappers::HbSubsetInputPtr input(hb_subset_input_create_or_fail());
+  {
+    hb_set_t* desired_codepoints = hb_subset_input_unicode_set(input.get());
+    HarfbuzzWrappers::HbSetPtr actual_codepoints(hb_set_create());
+    hb_face_collect_unicodes(font_face.get(), actual_codepoints.get());
+    std::string raw_codepoint;
+    while (std::cin >> raw_codepoint) {
+      auto codepoint = ParseCodepoint(raw_codepoint);
+      if (!codepoint) {
```
Comment (zanderso) [inline tools/font-subset/main.cc]: It looks like `ParsedCodepoint()` can't return 0. Should this be an assert?

File: tools/font-subset/main.cc [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+void Usage() {
+  std::cout << "Usage:" << std::endl;
+  std::cout << "font-subset <output.ttf> <input.ttf>" << std::endl;
+  std::cout << std::endl;
+  std::cout << "The output.ttf file will be overwritten if it exists already "
+               "and the subsetting operation succeeds."
+            << std::endl;
+  std::cout << "Codepoints should be specified on stdin, separated by spaces, "
```
Comment (zanderso) [inline tools/font-subset/main.cc]: This documentation looks sufficient to me.

File: tools/font-subset/test.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  (True,  '1.ttf', MATERIAL_TTF, [r'0xE003']),
+  (True,  '1.ttf', MATERIAL_TTF, [r'\uE003']),
+  (False, '1.ttf', MATERIAL_TTF, [r'57348']), # False because different codepoint
+  (True,  '2.ttf', MATERIAL_TTF, [r'0xE003', r'0xE004']),
+  (True,  '2.ttf', MATERIAL_TTF, [r'0xE003', r'0xE004', r'57347',]), # Duplicated codepoint
+  (True,  '3.ttf', MATERIAL_TTF, [r'0xE003', r'0xE004', r'0xE021',]),
+)
+
+FAIL_TESTS = [
+  ([FONT_SUBSET, 'output.ttf', 'does-not-exist.ttf'], ['1',]), # non-existant input font
```
Comment (zanderso) [inline tools/font-subset/test.py]: How about an empty string for a codepoint.

File: tools/font-subset/main.cc (hunk 3 above)
Comment (dnfield) [reply tools/font-subset/main.cc]: Ahh yes - at one point, I was letting this be a no-op, but I think that it's better to just hard exit when we don't knwo what's going on.

File: tools/font-subset/test.py (hunk 5 above)
Comment (dnfield) [reply tools/font-subset/test.py]: Updated to add that case and to make sure it fails.

Comment (zanderso) [reply tools/font-subset/test.py]: That isn't empty, it's a space ;-)

**Comment by dnfield** [comment]
Going to land this to try to kick CI - the depot_tools issue should be resolved now.

**Pull request merged**